### DISCUSSION TITLE
Does update return before its task is completed?

### DISCUSSION THREAD

**Discussion by epikgit** [ujjwalguptaofficial/JsStore#388] [Q&A] [answered]
I have a really strange behaviour with a particular JsStore update command.

This is my wrapper lib to save an svg image in an indexedDb table called WorkingImages.
```
db.setSuspendImage = async function (idImageRemote, currentSvg) {   
    await db.update({   
        in: 'WorkingImages',  
	    where: {idImageRemote: idImageRemote},  
	    set: {
                done: 1,  
		suspendedSvgMarkup: currentSvg,  
	    },  
	});  
} 
```

This is the where I'm calling it.

```
$('#btnSave').on('click', async function () {  
  await db.setSuspendImage(_CurrentIdImageRemote, $('#svgC svg').prop('outerHTML'));  
  document.location.href = "./#";  
});
```

I think it's clear what I'm doing here: I want to save the image markup with the wrapper and, when it finished, going to another page.
The strange thing is that this kind of approach doesn't work because I always find the image is not saved.

What I have to do to accomplish this simple task is this bestiality:

```
$('#btnSave').on('click', async function () {
	await db.setSuspendImage(_CurrentIdImageRemote, $('#svgC svg').prop('outerHTML'));
	setTimeout(() => {
		document.location.href = "./#";
	}, 1000);
});
```

With a 1 sec timeout everything works. 
It seems that the update method doesn't return its promise when it finishes but just some moments before the real end. Changing the page kill the js thread and the JsStore update.

Consider that the svg is really big (up to 8Mb).

JsStore version 4.8.4

**Comment by ujjwalguptaofficial** [comment]
I see, could you provide any reproducable example ?

**Comment by epikgit** [comment]
Unfortunately not because the code is subject to copyright and it's really complex to rewrite.

But if it could be useful I can add another important detail.
If I execute the code on a powerful PC, everything works well even without the artificial delay. Instead if I execute the same code on a mobile device, the problem arises. I added the settimeout to be sure it works with every HW.

IMHO it could be a sort of race condition and nobody has realized that before because, with a lot of computing power, everything works because the race is always won by JsStore.
My case is particular because I have powerless HW and the page is abandoned as soon as the update asserts it has finished.

**Reply by ujjwalguptaofficial** [reply]
make sense, i will check at the code side. Could you tell me what's tjhe format of your file, i mean string or blob ?

**Comment by epikgit** [comment]
It's a string extracted from the page markup.
To be clear I can have a situation like this:
```
<html>
<body>
...
<div id="svgC">
   <svg height="210" width="400" xmlns="http://www.w3.org/2000/svg">
    <path d="M150 5 L75 200 L225 200 Z" style="fill:none;stroke:green;stroke-width:3" />
  </svg> 
</div>
...
</body>
</html>
```

The only difference is that in my case the svg markup can weigh up to 8Mb.

**Reply by ujjwalguptaofficial** [reply]
I have found the issue, we update the data and return the result ASAP and does not wait for transaction completion. It works most of the time as data is small but in your case due to large data i guess you are having problem.

I am working on a option support to make it optional like wait for transaction or directly return the result.

Will explain more.

**Comment by epikgit** [comment]
Well in my opinion the solution shouldn't be as complicated as you are designing.

Since the update is an async operation you should always wait for a full promise completion. 
The decision to wait for the promise resolution or move forward immediately is on the user side.
So if I'm interested in the outcome I use await, otherwise I don't.

**Reply by ujjwalguptaofficial** [reply]
Actually it is like that only but from indexedb perspective - within one tx you can exeute multiple queries like select, update, remove etc.

In our case we are using only update and after that we dont need tx anymore, so after all update is done we return the result but actual tx is still open and waiting for any query , it will be closed when no further query is executed assumption can be for few milliseconds.

anyway now in the fix - i m waiting for tx completion by default, and a option is provided to return tx immediately. will update you once released.

**Comment by ujjwalguptaofficial** [comment, marked as answer]
@epikgit This issue has been fixed in v - 4.9.0 . Now update query by default wait for indexeddb tx completion which will make sure your data has been updated properly.

You don't need to change anything in code, just update the version to 4.9.0.

Let me know if you have any question, doubts or still seeing the same issue.

**Reply by ujjwalguptaofficial** [reply]
Also i have added a test case of updating a column with 10mb string size.
